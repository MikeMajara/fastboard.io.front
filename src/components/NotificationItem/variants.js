export default Object.freeze({
  info    : 'info',
  warning : 'warning',
  error   : 'error',
});
