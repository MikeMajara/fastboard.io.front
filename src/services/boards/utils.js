/**
 * Send boards event trough the network.
 *
 * @param {String} eventName Event key name.
 * @param {Object} data Data associated to the event.
 * @param {Function} ask Callback for server event acknowledgement.
 */
function send(eventName, data, ack) {
  this.socket.emit(eventName, data, ack);
}

/**
 * Check if the service is initialized.
 *
 * @return {Boolean} True if the service is initialized; false otherwise.
 */
function isInit() {
  return this.isInit;
}

export {
  isInit,
  send,
};
