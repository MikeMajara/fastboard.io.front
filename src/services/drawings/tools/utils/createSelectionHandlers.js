import { Group, Shape } from 'paper';

import { Log } from '#utils';
import { cursorTypes, selectionColorCode } from '#constants';
import store, { setSelectorCursorHover } from '#store';

import { bounds } from '../selector/constants';

const boundsArray = Object.values(bounds);
const bounds2cursorType = Object.freeze({
  [bounds.topLeft]     : cursorTypes.rotating,
  [bounds.topRight]    : cursorTypes.menu,
  [bounds.bottomLeft]  : cursorTypes.resizingSW,
  [bounds.bottomRight] : cursorTypes.resizingSE,
});

/**
 * Add selection handlers to the selected item.
 *
 * @param {Object} selectedItem Item to be added selection handlers to.
 * @param {Object} selectionLayer Layer that will contain the selection handlers.
 *
 * @returns {Object} Item - group with 4 handler items - from the selection handler layer.
 */
export default function createSelectionHandlers(selectedItem, selectionLayer) {
  Log.debug('Service : Drawings : Tools : Utils : createSelectionHandlers', { selectedItem });

  // Set selected item to show its selection state
  selectedItem.set({
    selected      : true,
    selectedColor : selectionColorCode,
    onMouseEnter  : () => store.dispatch(setSelectorCursorHover(cursorTypes.dragging)),
    onMouseLeave  : () => store.dispatch(setSelectorCursorHover()),
  });

  // Create bound handlers on the selection layer
  boundsArray.forEach((bound) => {
    selectionLayer.addChild(new Shape.Circle({
      name         : bound,
      center       : selectedItem.internalBounds[bound],
      radius       : 6,
      fillColor    : selectionColorCode,
      onMouseEnter : () => store.dispatch(setSelectorCursorHover(bounds2cursorType[bound])),
      onMouseLeave : () => store.dispatch(setSelectorCursorHover()),
    }));
  });

  // Create a group with the handlers to manipulate all of them at once
  selectionLayer.addChild(
    new Group({
      name     : 'selectionHandlers', // @todo to constants
      children : boundsArray.map(
        (bound) => selectionLayer.children[bound],
      ),
    }),
  );

  // Reproduce the selected item transformations on the handler items group
  selectionLayer.children.selectionHandlers.transform(selectedItem.matrix);

  // Except the scaling of the handler items themselves, which retain the original size
  selectionLayer.children.selectionHandlers.children.forEach((handlerItem) => {
    handlerItem.set({ scaling : 1 });
  });

  return selectionLayer.children.selectionHandlers;
}
