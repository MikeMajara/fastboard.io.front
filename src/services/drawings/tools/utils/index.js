export { default as createSelectionHandlers } from './createSelectionHandlers';
export { default as removeSelectionHandlers } from './removeSelectionHandlers';
export { default as resizeSelectionHandlers } from './resizeSelectionHandlers';
export { default as rotateSelectionHandlers } from './rotateSelectionHandlers';
export { default as translateSelectionHandlers } from './translateSelectionHandlers';
