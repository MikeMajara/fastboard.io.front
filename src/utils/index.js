export { default as Log } from './logger';
export { default as generateId } from './generateId';
export { default as point2net } from './point2net';
export { default as setPreventUnload } from './setPreventUnload';
export { default as throttle } from './throttle';
export { default as timeoutPromise } from './timeoutPromise';
export { default as triggerDownload } from './triggerDownload';
export { default as triggerUpload } from './triggerUpload';
export { default as validBoardId } from './validBoardId';
