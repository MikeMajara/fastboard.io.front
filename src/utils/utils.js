/**
 * Throttles a function call.
 *
 * @param {Function} callback Function to be throttled.
 * @param {Integer} delay Throttle delay in milliseconds.
 *
 * @return {Function} Function that throttles the execution of the callback.
 */
export const throttle = (callback, delay = 0) => {
  let previousTime = new Date().getTime();

  return function throttledCallback(args) {
    const currentTime = new Date().getTime();

    if ((currentTime - previousTime) >= delay) {
      previousTime = currentTime;
      callback(...args);
    }
  };
};

/**
 * Creates a Promise with a rejection timeout.
 *
 * @param {Function} callback Function for the promise. Receives res and rej functions.
 * @param {Integer} timeout Timeout for the promise to be rejected in milliseconds.
 *
 * @return {Promise} Promise with rejection timeout.
 */
export const timeoutPromise = (callback, timeout) => new Promise(
  (res, rej) => {
    const timeoutId = setTimeout(() => rej(new Error()), timeout);

    callback(
      (...args) => {
        clearTimeout(timeoutId);
        res(...args);
      },
      (...args) => {
        clearTimeout(timeoutId);
        rej(new Error(...args));
      },
    );
  },
);
