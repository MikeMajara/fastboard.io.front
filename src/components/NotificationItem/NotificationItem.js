import React from 'react';
import propTypes from 'prop-types';
import { useTranslation } from 'react-i18next';

import { notificationTypes } from '#constants';

import variants from './variants';
import SWrapper from './styled';

const testId = 'notificationitem-component';

// Notification item style varies depending on type
const type2variant = Object.freeze({
  [notificationTypes.userJoined]       : variants.info,
  [notificationTypes.userLeft]         : variants.info,
  [notificationTypes.userNameSet]      : variants.info,
  [notificationTypes.boardNameSet]     : variants.info,
  [notificationTypes.boardCleared]     : variants.info,
  [notificationTypes.boardExportError] : variants.error,
});

const NotificationItem = ({ type, data }) => {
  const { t } = useTranslation('board');

  return (
    <SWrapper variant={type2variant[type]} data-testid={testId}>
      {t(`notifications.${type}`, { ...data })}
    </SWrapper>
  );
};

NotificationItem.defaultProps = {
  data : undefined,
};

NotificationItem.propTypes = {
  type : propTypes.oneOf(Object.values(notificationTypes)).isRequired,
  data : propTypes.shape({}),
};

export default NotificationItem;
