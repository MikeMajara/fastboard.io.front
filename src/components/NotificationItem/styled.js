import styled, { css } from '@xstyled/styled-components';
import { variant } from '@xstyled/system';

import variants from './variants';

export default styled.div`
  @keyframes appear {
    from {
      opacity: 0;
      transform: translate(0, 100%);
    }
  }

  display: flex;
  width: notificationWidth;
  justify-content: center;
  padding: default;
  animation: appear .2s ease-in-out;
  border-radius: default;
  box-shadow: default;
  font-size: sm;

  ${variant({
    default  : 'generic',
    prop     : 'variant',
    variants : {
      [variants.info] : css`
        background-color: rgba(255, 255, 255, .94);
        color: fg;
      `,
      [variants.warning] : css`
        background-color: rgba(255, 170, 60, .94);
        color: #fff;
      `,
      [variants.error] : css`
        background-color: rgba(255, 80, 80, .94);
        color: #fff;
      `,
    },
  })}
`;
