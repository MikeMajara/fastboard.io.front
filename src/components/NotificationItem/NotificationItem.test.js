import React from 'react';

import { render } from '#devTools/tests/utils';
import { notificationTypes } from '#constants';

import NotificationItem from './NotificationItem';

describe('Component : NotificationItem', () => {
  test('Is visible with notification content', () => {
    const { queryByText } = render(
      <NotificationItem
        type={notificationTypes.userJoined}
        data={{}}
      />,
    );

    expect(queryByText(`notifications.${notificationTypes.userJoined}`)).toBeInTheDocument();
  });

  test('Snapshot', () => {
    const { asFragment } = render(
      <NotificationItem
        type={notificationTypes.userJoined}
        data={{}}
      />,
    );

    const componentRender = asFragment();
    expect(componentRender).toMatchSnapshot();
  });
});
