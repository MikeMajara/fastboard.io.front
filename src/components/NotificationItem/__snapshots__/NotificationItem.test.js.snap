// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component : NotificationItem Snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-pscky iSRUFQ"
    data-testid="notificationitem-component"
  >
    notifications.userJoined
  </div>
</DocumentFragment>
`;
